otlp/traces_direct.go

```diff
@@ -0,0 +1,1709 @@
+		if err != nil {
+			return err
+		}
+
+		switch fieldNum {
+		case 1: // attributes
+			slice, err := decodeWireType2(data, &iNdEx, l, wireType)
+			if err != nil {
+				return err
+			}
+
+			// Parse KeyValue
+			err = unmarshalKeyValue(ctx, slice, attrs, 0)
+			if err != nil {
+				return err
+			}
+
+		default:
+			if err := skipField(data, &iNdEx, preIndex, l); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// parseKeyValue parses a KeyValue message and returns the key and raw value bytes
+func parseKeyValue(data []byte) ([]byte, []byte, error) {
+	var key []byte
+	var valueBytes []byte
+
+	l := len(data)
+	iNdEx := 0
+
+	for iNdEx < l {
+		preIndex := iNdEx
+		fieldNum, wireType, err := decodeField(data, &iNdEx)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		switch fieldNum {
+		case 1: // key
+			key, err = decodeWireType2(data, &iNdEx, l, wireType)
+			if err != nil {
+				return nil, nil, err
+			}
+
+		case 2: // value
+			valueBytes, err = decodeWireType2(data, &iNdEx, l, wireType)
+			if err != nil {
+				return nil, nil, err
+			}
+
+		default:
+			if err := skipField(data, &iNdEx, preIndex, l); err != nil {
+				return nil, nil, err
+			}
+		}
+	}
+
+	return key, valueBytes, nil
+}
+
+// processValueDirect handles a value recursively, tracking depth for proper flattening
+func processValueDirect(ctx context.Context, key []byte, value any, attrs *msgpAttributes, depth int) error {
+	var err error
+
+	// Handle different value types to match legacy behavior
+	switch v := value.(type) {
+	case []byte:
+		// Bytes are JSON encoded - match the legacy behavior
+		husky.AddTelemetryAttribute(ctx, "received_bytes_attr_type", true)
+		err = attrs.addAny(key, addAttributeToMapAsJsonDirect(v))
+		if err != nil {
+			return err
+		}
+	case []any:
+		// Arrays are JSON encoded
+		husky.AddTelemetryAttribute(ctx, "received_array_attr_type", true)
+		err = attrs.addAny(key, addAttributeToMapAsJsonDirect(v))
+		if err != nil {
+			return err
+		}
+	case map[string]any:
+		// Kvlists are flattened with dot notation
+		husky.AddTelemetryAttributes(ctx, map[string]any{
+			"received_kvlist_attr_type": true,
+			"kvlist_max_depth":          depth,
+		})
+		if depth < maxDepth {
+			// Flatten the kvlist
+			for k, v := range v {
+				flatKey := append(key, '.')
+				flatKey = append(flatKey, k...)
+
+				// Process the nested value recursively
+				err = processValueDirect(ctx, flatKey, v, attrs, depth+1)
+				if err != nil {
+					return err
+				}
+			}
+		} else {
+			// Max depth exceeded, JSON encode the whole thing
+			err = attrs.addAny(key, addAttributeToMapAsJsonDirect(v))
+			if err != nil {
+				return err
+			}
+		}
+	default:
+		// Simple types - just encode directly
+		err = attrs.addAny(key, v)
+		if err != nil {
+			return err
+		}
+	}
+
+	// If this is the service name, note it.
+	if asStr, ok := value.(string); ok && bytes.Equal(key, []byte("service.name")) {
```

Comment:
Is it expected to use the value find from the deeply nested map here as the service name?
I believe the original behavior would have changed the key to "parent.service.name" depending on the level